doc/sips/remote-metadata/README.md

```diff
@@ -0,0 +1,310 @@
+first sector). A record of type MetadataCount is added to the new
+sector. When such a record is parsed, the parser gets the total
+number of metadata sectors. Thus, in the diagram above **Metadata N**
+sector stores the oldest metadata and **Metadata 1** - the youngest.
+
+How to parse metadata:
+
+```
+for {
+    if end of sector {
+        break
+    }
+    read length
+    if length == 0 {
+        break
+    }
+    read change from next length bytes
+    if change.type == MetadataCount {
+        read metadata sectors from change.count to 1 first
+    }
+}
+```
+
+Sia supports atomic updates of multiple sectors. To change anything,
+renter uploads data sectors and corresponding metadata changes in the
+same request. Metadata changes are small because they only change
+the affected bytes of the first sector.
+
+If a sector of a file is discarded or the whole file is removed, then
+the corresponding sector of the host becomes free. When a new sector
+is needed on the host, free sectors are used if they are available.
+If a sector of a file is overwritten, a new version is written to new
+(or free) sector, then parity sectors are written to other hosts,
+after that the old sector and old parity sectors become free.
+`Write` and `Discard` records have `revision` field that is used
+to determine which information about the sector is fresh.
+
+Before renewing a contract, the renter completely removes free sectors
+from the contract by sending Delete RPC and rewrites all the metadata.
+Metadata rewriting is needed because indices of sectors are shifted
+when free data sectors are removed. It is also a good time to skip
+unneeded parts of metadata such as writes that were overwritten
+subsequently.
+
+Parity sectors are added to a group of sectors stored on different
+servers and are uploaded to a non-overlapping set of servers. Metadata
+of a parity sector contains corresponding `Write` records of all data
+sectors involved. If the host holding the data sector is down,
+the renter can recover all data and metadata from parity sectors.
+
+A file is identified with unique integer identifier (e.g. counter).
+After a file is removed, its id is never reused.
+
+### Encryption
+
+Thanks to Merkle root and sector IDs, we don't have to authenticate
+data: if it is tampered by the host (or when transferred to renter),
+the sector hash will be different. So we only need to encrypt it to
+keep safe.
+
+A contract has the following types of sectors:
+
+  * metadata
+  * data
+  * parity
+  * free
+
+A free sector does not hold usable data. When it was not free (i.e.
+was used for something) it was encrypted and that data is still there.
+
+Sectors are encrypted with a stream cipher with random access (i.e.
+which can produce bytes for XOR in the arbitrary location of the
+sequence). One contract corresponds to one sequence (i.e. one
+encryption key). The cipher has large space for its pseudorandom
```

Comment:
Encrypting all of the data in a contract with the same key causes problems for file-sharing. A contract can contain data from multiple files, so sharing the decryption key for one file would mean sharing the key for all other files.

Reply:
I overlooked file sharing.

We should use a separate key for each file. Metadata will be encrypted with a per-host key. The only thing to take into account is that if the file is shared (with the key) and somebody knows the list of IDs of hosts involved in storing the contracts (e.g. hosts themselves), then they can decrypt the previous data from the file. It can be not OK for some users who want to publish the final version of a file but keep preliminary versions secretly.

More issues are added by parity sectors. If all data sectors in a group are from the same file, then parity sectors can be encrypted with the same key. But if data sectors of a group originate from multiple files, we have to make sure we don't leak other files when the user shares just one file involved in the group. One idea is to add parity sectors after encrypting the data, so people who obtain the shared file can only decrypt an encrypted version of data from other files. What do you think?